modules/apps/app-builder/app-builder-test/src/testIntegration/java/com/liferay/app/builder/deploy/test/AppDeployerTest.java

```diff
@@ -0,0 +1,217 @@
+		throws Exception {
+
+		AppBuilderApp appBuilderApp = _createAppBuilderApp();
+
+		JSONObject settingsJSONObject = JSONFactoryUtil.createJSONObject(
+			JSONUtil.put(
+				"scope", JSONUtil.put("applications_menu.applications")
+			).toJSONString());
+
+		_appBuilderAppDeploymentLocalService.addAppBuilderAppDeployment(
+			appBuilderApp.getAppBuilderAppId(),
+			settingsJSONObject.toJSONString(), "productMenu");
+
+		_deployAndUndeployAppBuilderApp(
+			appBuilderApp, "applications_menu.applications");
+	}
+
+	@Test
+	public void testProductMenuDeployerDeployInApplicationsMenuAndSites()
+		throws Exception {
+
+		AppBuilderApp appBuilderApp = _createAppBuilderApp();
+
+		JSONArray jsonArray = JSONFactoryUtil.createJSONArray();
+
+		JSONObject siteAndApplicationsJSONObject =
+			JSONFactoryUtil.createJSONObject(
+				JSONUtil.put(
+					"scope",
+					jsonArray.put(
+						"applications_menu.applications"
+					).put(
+						"site_administration.content"
+					)
+				).put(
+					"siteIds", JSONUtil.put(-1)
+				).toJSONString());
+
+		_appBuilderAppDeploymentLocalService.addAppBuilderAppDeployment(
+			appBuilderApp.getAppBuilderAppId(),
+			siteAndApplicationsJSONObject.toJSONString(), "productMenu");
+
+		_deployAndUndeployAppBuilderApp(
+			appBuilderApp, "applications_menu.applications",
+			"site_administration.content");
+	}
+
+	@Test
+	public void testProductMenuDeployerDeployInSites() throws Exception {
+		AppBuilderApp appBuilderApp = _createAppBuilderApp();
+
+		JSONObject settingsJSONObject = JSONFactoryUtil.createJSONObject(
+			JSONUtil.put(
+				"scope", JSONUtil.put("site_administration.content")
+			).put(
+				"siteIds", JSONUtil.put()
+			).toJSONString());
+
+		_appBuilderAppDeploymentLocalService.addAppBuilderAppDeployment(
+			appBuilderApp.getAppBuilderAppId(),
+			settingsJSONObject.toJSONString(), "productMenu");
+
+		_deployAndUndeployAppBuilderApp(
+			appBuilderApp, "site_administration.content");
+	}
+
+	private AppBuilderApp _createAppBuilderApp() throws Exception {
+		return _appBuilderAppLocalService.addAppBuilderApp(
+			TestPropsValues.getGroupId(), TestPropsValues.getCompanyId(),
+			TestPropsValues.getUserId(), true, RandomTestUtil.nextLong(),
+			RandomTestUtil.nextLong(), RandomTestUtil.nextLong(),
+			RandomTestUtil.nextLong(), RandomTestUtil.randomLocaleStringMap(),
+			RandomTestUtil.randomString());
+	}
+
+	private void _deployAndUndeployAppBuilderApp(
+			AppBuilderApp appBuilderApp, String... expectedPanelCategoryKeys)
+		throws Exception {
+
+		AppDeployer appDeployer = _appDeployerTracker.getAppDeployer(
+			"productMenu");
+
+		appDeployer.deploy(appBuilderApp.getAppBuilderAppId());
+
+		Map<Long, ServiceRegistration<?>[]> serviceRegistrationsMap =
+			ReflectionTestUtil.getFieldValue(
+				appDeployer, "_serviceRegistrations");
+
+		ServiceRegistration<?>[] serviceRegistrations =
+			serviceRegistrationsMap.get(appBuilderApp.getAppBuilderAppId());
```

Comment:
How about bringing this closer to the for loop? (as used) - _I'm still a little bit confused about the burrito style_ -

Reply:
the order is alrigth. Ping me so i can clarify it ^^